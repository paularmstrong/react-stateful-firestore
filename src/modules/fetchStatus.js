export const FetchStatus = {
  LOADING: 'loading',
  LOADED: 'loaded',
  FAILED: 'failed',
  NONE: 'none'
};
