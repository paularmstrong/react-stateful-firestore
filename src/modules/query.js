// @flow
import type { FieldPath, Query } from '@firebase/firestore';
import type { StoreState } from '../reducers';

type Operation = { name: '<' | '<=' | '==' | '>=' | '>' };
type Value = { internalValue: any };
type RelationalFilter = {
  field: FieldPath,
  op: Operation,
  value: Value
};

const emptyArray = [];

// $FlowFixMe Undocumented key/value
const pathToString = (path: FieldPath) => path.segments.join('/');

const filterToString = (filter: RelationalFilter) => {
  const { field, op, value } = filter;
  return `${pathToString(field)}${op.name}${value.toString()}`;
};

export const getQueryId = (query: Query): string => {
  // $FlowFixMe Undocumented key/value
  if (query.id && query.id === query.path) {
    // $FlowFixMe Undocumented key/value
    return query.id;
  }

  if (query.path && typeof query.path === 'string') {
    // $FlowFixMe Undocumented key/value
    return query.path;
  }

  if (query._query) {
    // $FlowFixMe Grabbing private value
    const { filters, limit, path } = query._query;
    const filterString = filters.map(filterToString).join('|');
    return `${pathToString(path)}${filterString ? `:${filterString}` : ''}${limit ? `:${limit}` : ''}`;
  }

  throw new Error('Unknown query type');
};

export const getQueryPath = (query: Query): string => {
  if (query.path) {
    // $FlowFixMe Undocumented key/value
    return query.path;
  }

  if (query._query) {
    // $FlowFixMe Grabbing private value
    const { path } = query._query;
    return pathToString(path);
  }

  throw new Error('Unknown query type');
};

export const getCollectionQueryPath = (query: Query): string => {
  const fullPath = getQueryPath(query);
  const pathParts = fullPath.split('/');
  return pathParts.length % 2 ? fullPath : fullPath.replace(/(\/[^/]+)$/, '');
};

const _getValueFromFieldPath = (doc, fieldPath: FieldPath) => {
  let value = doc;
  // $FlowFixMe Undocumented key/value
  for (let i = 0; i < fieldPath.segments.length; i++) {
    // $FlowFixMe Undocumented key/value
    const segment = fieldPath.segments[i];
    value = value && typeof value === 'object' ? value[segment] : undefined;
    if (typeof value === undefined) {
      return undefined;
    }
  }
  return value;
};

export const getDocumentIdsForQuery = (query: Query, state: StoreState): Array<string> => {
  const collectionPath = getCollectionQueryPath(query);
  const queryPath = getQueryPath(query);
  const pathParts = queryPath.split('/');
  const documents = state.collections[collectionPath];
  if (!(pathParts.length % 2)) {
    const documentId = pathParts[pathParts.length - 1];
    return Object.keys(documents).indexOf(documentId) >= 0 ? [documentId] : [];
  }

  if (query._query) {
    // $FlowFixMe Undocumented key/value
    const { filters, limit, orderBy } = query._query;
    let docs = Object.values(documents).filter((doc) => {
      return filters.every(({ field, op, value }) => {
        const docValue = _getValueFromFieldPath(doc, field);
        const { internalValue } = value;
        const normValue =
          typeof internalValue === 'object'
            ? internalValue.seconds ? new Date(internalValue.seconds) : internalValue
            : internalValue;
        switch (op.name) {
          case '<':
            return docValue < normValue;
          case '<=':
            return docValue <= normValue;
          case '==':
            return docValue == normValue;
          case '>':
            return docValue > normValue;
          case '>=':
            return docValue >= normValue;
          default:
            return true;
        }
      });
    });

    if (orderBy && orderBy.length > 0) {
      orderBy.forEach((order) => {
        const { dir: { name: dir }, field } = order;
        docs = docs.sort((a, b) => {
          const aValue = _getValueFromFieldPath(a, field);
          const bValue = _getValueFromFieldPath(b, field);
          if (!aValue || !bValue) {
            return 0;
          }
          if (typeof aValue === 'string' && typeof bValue === 'string') {
            return dir === 'asc' ? aValue.localeCompare(bValue) : bValue.localeCompare(aValue);
          }
          if (aValue instanceof Date && bValue instanceof Date) {
            return dir === 'asc' ? aValue.valueOf() - bValue.valueOf() : bValue.valueOf() - aValue.valueOf();
          }
          if (typeof aValue === 'number' && typeof bValue === 'number') {
            return dir === 'asc' ? aValue - bValue : bValue - aValue;
          }
          return 0;
        });
      });
    }

    return docs
      .map((doc) => (typeof doc === 'object' && doc && typeof doc.id === 'string' ? doc.id : undefined))
      .filter(Boolean)
      .slice(0, limit);
  }
  return emptyArray;
};
