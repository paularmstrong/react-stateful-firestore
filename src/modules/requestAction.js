export const RequestAction = {
  REQUEST: 'request',
  SUCCESS: 'success',
  FAILURE: 'failure'
};
