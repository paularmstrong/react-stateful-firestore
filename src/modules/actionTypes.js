// @flow
import { RequestAction } from '../modules/requestAction';

const NAMESPACE = 'firestore';

export const createActionType = (action: string) => `${NAMESPACE}/${action}`;

type RequestActionTypes = {
  REQUEST: string,
  SUCCESS: string,
  FAILURE: string
};

export const createRequestActionTypes = (request: string): RequestActionTypes => {
  return Object.keys(RequestAction).reduce(
    (memo: RequestActionTypes, key: $Keys<RequestActionTypes>) => {
      memo[key] = createActionType(`${request}/${key}`);
      return memo;
    },
    { REQUEST: '', SUCCESS: '', FAILURE: '' }
  );
};
