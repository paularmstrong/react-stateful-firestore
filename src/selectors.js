// @flow
import { createSelector } from 'reselect';
import { FetchStatus } from './modules/fetchStatus';
import { getCollectionQueryPath, getQueryId, getQueryPath } from './modules/query';
import { addListener, addQuery } from './actions';

import type { Auth } from 'firebase/auth';
import type { Query } from 'firebase/firestore';
import type { Store } from 'redux';
import type { StoreState } from './reducers';
import type { QueryState } from './reducers/queries';

type Props = {};

const emptyArray = [];
const emptyObject = {};

const singleUndefined = { fetchStatus: FetchStatus.NONE, doc: undefined };
const collectionUndefiend = { fetchStatus: FetchStatus.NONE, docs: emptyArray };

const selectQueries = (state: StoreState) => state.queries;

export const initSelect = (store: Store<*, *, *>) => (query: Query) => {
  const queryId = getQueryId(query);
  const queryPath = getQueryPath(query);
  const collectionQueryPath = getCollectionQueryPath(query);
  const isDocument = collectionQueryPath !== queryPath;

  const selectStoreQuery = (state: StoreState) => state.queries[queryId];
  const selectCollection = (state: StoreState) => state.collections[collectionQueryPath];

  const selector = createSelector([selectStoreQuery, selectCollection], (storeQuery, collection) => {
    if (storeQuery && collection) {
      const { documentIds, fetchStatus } = storeQuery;
      if (documentIds) {
        const docs = documentIds.map((id: string) => collection[id]).filter(Boolean);
        return isDocument ? { fetchStatus, doc: docs[0] } : { fetchStatus, docs };
      }
    }
    return isDocument ? singleUndefined : collectionUndefiend;
  });

  return () => {
    store.dispatch(addQuery(query));
    store.dispatch(addListener(query));
    return selector;
  };
};

export const initSelectAuth = (auth: Auth, userCollection?: string) => {
  const loggedOut = { fetchStatus: FetchStatus.NONE, doc: undefined };
  const selectCollectionName = () => userCollection;
  const selectUid = (state: StoreState) => state.auth.uid;
  const selectStoreQuery = createSelector(
    [selectCollectionName, selectUid, selectQueries],
    (uid, queries) => (userCollection ? queries[`auth|${userCollection}/${uid}`] : undefined)
  );
  const selectUsersCollection = (state: StoreState) => (userCollection ? state.collections[userCollection] : undefined);
  const selectUserData = createSelector(
    [selectUid, selectUsersCollection],
    (uid, users) => (users ? users[uid] : undefined)
  );
  const selectCurrentUser = (state: StoreState) =>
    auth.currentUser ? JSON.stringify(auth.currentUser.toJSON()) : undefined;

  const selector = createSelector([selectUid, selectCurrentUser, selectStoreQuery, selectUserData], (
    uid: string,
    currentUser?: string, // used for selector cache busting. Use auth.currentUser with connectAuth.
    storeQuery?: QueryState,
    doc: {}
  ) => {
    if (!uid) {
      return loggedOut;
    }
    const fetchStatus = storeQuery ? storeQuery.fetchStatus : FetchStatus.LOADED;
    return {
      fetchStatus,
      doc: userCollection ? doc : {} // use a new object if no userCollection to cache bust
    };
  });

  return selector;
};
