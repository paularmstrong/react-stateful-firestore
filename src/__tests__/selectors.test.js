import configureStore from 'redux-mock-store';
import { FetchStatus } from '../modules/fetchStatus';
import thunk from 'redux-thunk';
import { initSelect, initSelectAuth } from '../selectors';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const taco = { id: '123', type: 'delicious' };
const mockState = {
  auth: { uid: '123' },
  collections: { tacos: { [taco.id]: taco } },
  queries: { 'tacos/123': { documentIds: ['123'], fetchStatus: FetchStatus.LOADED } },
  listeners: {}
};
window.requestIdleCallback = jest.fn((cb) => cb());

describe('selectors', () => {
  const query = { id: 'tacos', path: 'tacos', get: () => Promise.resolve(), onSnapshot: jest.fn() };
  let store;
  let selector;

  beforeEach(() => {
    store = mockStore(mockState);
    selector = initSelect(store)(query);
  });

  describe('select', () => {
    test('dispatches adding the query and listener', () => {
      selector();
      expect(store.getActions()).toMatchSnapshot();
    });

    test('returns fetchStatus and user doc', () => {
      const selector = initSelect(store)({ ...query, id: `tacos/${taco.id}`, path: `tacos/${taco.id}` });
      const selectData = selector();
      expect(selectData(mockState)).toMatchSnapshot();
    });

    test('memoizes the response', () => {
      const selector = initSelect(store)({ ...query, id: `tacos/${taco.id}`, path: `tacos/${taco.id}` });
      const selectData = selector();
      const firstState = { ...mockState, collections: { ...mockState.collections, foobar: {} } };
      const secondState = { ...mockState, collections: { ...firstState.collections, foobar: { '456': {} } } };
      expect(selectData(firstState)).toBe(selectData(secondState));
    });
  });

  describe('selectAuth', () => {
    test('returns fetchStatus and user doc', () => {
      let currentUserJSON = { uid: '123' };
      const selectData = initSelectAuth({ currentUser: { uid: '123', toJSON: () => currentUserJSON } }, 'tacos')();
      expect(selectData(mockState)).toMatchSnapshot();
    });

    test('memoizes the response', () => {
      let currentUserJSON = { uid: '123' };
      const selectData = initSelectAuth({ currentUser: { uid: '123', toJSON: () => currentUserJSON } }, 'tacos')();
      const firstState = { ...mockState, collections: { ...mockState.collections, foobar: {} } };
      const secondState = { ...mockState, collections: { ...mockState.collections, foobar: { '456': {} } } };
      expect(selectData(firstState)).toBe(selectData(secondState));
    });

    test('current user busts memoization', () => {
      const toJSON = jest.fn(() => {
        uid: '123';
      });
      const selectData = initSelectAuth({ currentUser: { uid: '123', toJSON } }, 'tacos')();
      const firstData = selectData(mockState);
      const secondState = { ...mockState };
      toJSON.mockReturnValueOnce({ uid: '123', foo: 'bar' });
      expect(firstData).not.toBe(selectData(secondState));
    });
  });
});
