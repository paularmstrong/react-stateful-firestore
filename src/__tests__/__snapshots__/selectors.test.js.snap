// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors select dispatches adding the query and listener 1`] = `
Array [
  Object {
    "meta": Object {
      "queryId": "tacos",
    },
    "payload": Object {
      "get": [Function],
      "id": "tacos",
      "onSnapshot": [MockFunction] {
        "calls": Array [
          Array [
            [Function],
          ],
        ],
      },
      "path": "tacos",
    },
    "type": "firestore/queries/REQUEST",
  },
  Object {
    "meta": Object {
      "queryId": "tacos",
    },
    "payload": undefined,
    "type": "firestore/listeners/ADD",
  },
]
`;

exports[`selectors select returns fetchStatus and user doc 1`] = `
Object {
  "doc": Object {
    "id": "123",
    "type": "delicious",
  },
  "fetchStatus": "loaded",
}
`;

exports[`selectors selectAuth if no userCollection provided, returns empty object 1`] = `
Object {
  "doc": Object {},
  "fetchStatus": "loaded",
}
`;

exports[`selectors selectAuth returns fetchStatus and user doc 1`] = `
Object {
  "doc": Object {
    "id": "123",
    "type": "delicious",
  },
  "fetchStatus": "loaded",
}
`;
